### ISSUE TITLE
Have locked in versions for dependencies 

### ISSUE THREAD

**Issue by BioPhoton** [code-pushup/cli#500]
Having inter-package dependencies like "@code-pushup/models": "*" in our package.jsons is a problem, they should all be locked into the same version. Otherwise only direct dependencies like cli get updated and indirect dependencies stay behind.

Todo:
- set fixed version for all packages 
- Avoid “*” versions with eslint or a plugin 

Research:
- would it help to replace [jscutlery/semver](https://github.com/jscutlery/semver) with new [`nx release`](https://nx.dev/nx-api/nx/documents/release)?

**Comment by getlarge** [comment]
I am willing to take on this task if it helps.

**Comment by getlarge** [comment]
A [generator ](https://github.com/jscutlery/semver/pull/761/files) has been created to ease migration to `nx release`.
I should probably start with this.

**Comment by getlarge** [comment]
@BioPhoton status update:
I just opened a draft PR with ongoing progress and need some help.
To use some needed features from the `nx release` CLI, I upgraded to `nx` 17.3.2 (the last 17.x), which required upgrading `vite` (^5.0.0) and `vitest` (^1.0.0), and unfortunately, the migration to `vitest` is not straightforward.

- When running the unit-test target for `core` the test run appears to be stuck until timed out, with no error displayed even when using the `--verbose` flag.
- When debugging `vitest` run by the Nx executor, i could log the following error emitted by the vitest worker:
```json
{
          "stack": "Error: [vitest-worker]: Timeout calling \"fetch\" with \"[\"/Users/edouard/Dev/code-pushup/cli/packages/core/src/lib/implementation/execute-plugin.unit.test.ts\",\"ssr\"]\"\n    at Object.onTimeoutError (file:///Users/edouard/Dev/code-pushup/cli/node_modules/vitest/dist/vendor/rpc.joBhAkyK.js:61:15)\n    at Timeout._onTimeout (file:///Users/edouard/Dev/code-pushup/cli/node_modules/vitest/dist/vendor/index.8bPxjt7g.js:39:41)\n    at listOnTimeout (node:internal/timers:569:17)\n    at processTimers (node:internal/timers:512:7)",
          "message": "[vitest-worker]: Timeout calling \"fetch\" with \"[\"/Users/edouard/Dev/code-pushup/cli/packages/core/src/lib/implementation/execute-plugin.unit.test.ts\",\"ssr\"]\"",
          "stackStr": "Error: [vitest-worker]: Timeout calling \"fetch\" with \"[\"/Users/edouard/Dev/code-pushup/cli/packages/core/src/lib/implementation/execute-plugin.unit.test.ts\",\"ssr\"]\"\n    at Object.onTimeoutError (file:///Users/edouard/Dev/code-pushup/cli/node_modules/vitest/dist/vendor/rpc.joBhAkyK.js:61:15)\n    at Timeout._onTimeout (file:///Users/edouard/Dev/code-pushup/cli/node_modules/vitest/dist/vendor/index.8bPxjt7g.js:39:41)\n    at listOnTimeout (node:internal/timers:569:17)\n    at processTimers (node:internal/timers:512:7)",
          "nameStr": "Error",
          "expected": "undefined",
          "actual": "undefined",
          "constructor": "Function<Error>",
          "name": "Error",
          "toString": "Function<toString>"
        }
```
All test suites result in the same error.

The cherry on top: the process is left hanging and never exits appropriately.

Since that is my first experience with `vite` / `vitest`, I would appreciate some input to progress.

**Comment by getlarge** [comment]
After reading this [vitest issue](https://github.com/vitest-dev/vitest/issues/3077) it seemed like fetch implementation in Node.JS might also be responsible for the hanging process, check this [undici issue](https://github.com/nodejs/undici/issues/2026).

So I tried to rerun the tests on a single thread. Result: the process is not left hanging anymore.

```ts
export default defineConfig({
  cacheDir: '../../node_modules/.vite/core',
  plugins: [
    nxViteTsPaths({
      debug: true,
    }),
  ],
  test: {
    name: 'core',
    globals: true,
    cache: {
      dir: '../../node_modules/.vitest/core',
    },
    pool: 'threads',
    poolOptions: {
      threads: {
        singleThread: true,
      },
    },
    coverage: {
      reporter: ['lcov'],
    },
    environment: 'node',
    include: ['src/**/*.unit.test.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
    globalSetup: ['../../global-setup.ts'],
    setupFiles: [
      '../../testing-utils/src/lib/setup/fs.mock.ts',
      '../../testing-utils/src/lib/setup/git.mock.ts',
      '../../testing-utils/src/lib/setup/console.mock.ts',
      '../../testing-utils/src/lib/setup/reset.mocks.ts',
      '../../testing-utils/src/lib/setup/portal-client.mock.ts',
    ],
    workspace: './vitest.workspace.ts',
  },
});
```

**Comment by getlarge** [comment]
A vite configuration that seems to work

```ts
/// <reference types="vitest" />
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';
import { URL } from 'node:url';
import { defineConfig } from 'vite';

export default defineConfig({
  cacheDir: '../../node_modules/.vite/core',
  plugins: [
    nxViteTsPaths({
      debug: true,
    }),
  ],
  resolve: {
    alias: [
      {
        find: '@code-pushup/testing-utils',
        replacement: '../../testing-utils/src',
      },
    ],
  },
  test: {
    name: 'core',
    globals: true,
    cache: {
      dir: '../../node_modules/.vitest/core',
    },
    alias: [
      {
        find: '@code-pushup/testing-utils',
        replacement: new URL('../../testing-utils/src', import.meta.url)
          .pathname,
      },
      {
        find: '@code-pushup/models',
        replacement: new URL('../models/src', import.meta.url).pathname,
      },
      {
        find: '@code-pushup/utils',
        replacement: new URL('../utils/src', import.meta.url).pathname,
      },
    ],
    pool: 'threads',
    poolOptions: {
      vmThreads: {
        singleThread: true,
      },
    },
    coverage: {
      reporter: ['lcov'],
      enabled: false,
    },
    environment: 'node',
    include: ['src/**/*.unit.test.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],
    globalSetup: ['../../global-setup.ts'],
    setupFiles: [
      '../../testing-utils/src/lib/setup/fs.mock.ts',
      '../../testing-utils/src/lib/setup/git.mock.ts',
      '../../testing-utils/src/lib/setup/console.mock.ts',
      '../../testing-utils/src/lib/setup/reset.mocks.ts',
      '../../testing-utils/src/lib/setup/portal-client.mock.ts',
    ],
    workspace: './vitest.workspace.ts',
  },
});

```

I would expect `nxViteTsPaths` to resolve those TS config paths...

**Comment by getlarge** [comment]
Still unable to make TS config path resolution working with Vitest, so i have been digging into issues and articles that seems related:
- https://github.com/nrwl/nx/issues/18492
- https://github.com/nrwl/nx/issues/19349
- https://github.com/nrwl/nx/issues/21030
- https://www.wolff.fun/path-mapping-typescript-vite/

When reading this particular [comment](https://github.com/nrwl/nx/issues/18492#issuecomment-1691942830) I thought that Vitest could be confused that the scope `@code-pushup` is used by some actual (external) packages and would try to resolve the local packages path (such as `@code-pushup/models`,  `@code-pushup/testing-utils`, …) from the node_modules, leading to the `Failed to load url @code-pushup/testing-utils` error message. 

If anyone has a clue what could go wrong, please share your knowledge :D
Otherwise, i will keep on reading and will venture into debugging how files are resolved by Vite.

**Comment by getlarge** [comment]
In the latest episode...
@BioPhoton joined to participate in a debug session, and we noticed the following:
- when running `npx vitest --config=vite.config.unit.ts` from the package directory, TS config paths are resolved thanks to `nxViteTsPaths`
- when patching `@nx/vite:test` executor to change dir before calling [`startVitest`](https://vitest.dev/advanced/api#startvitest) and reflect the path in the vitest options, `vite` resolve the TS paths correctly.

This is what the (ugly) patch looks like in `node_modules/@nx/vite/src/executors/test/vitest.impl.js`
```diff
// ...
+    resolvedOptions.root = '.';
+    resolvedOptions.config = path.basename(resolvedOptions.configFile);
+    process.chdir(path.join(process.cwd(), projectRoot));
    const ctx = await startVitest((_resolvedOptions_mode = resolvedOptions['mode']) != null ? _resolvedOptions_mode : 'test', cliFilters, resolvedOptions);
// ...
```

I still don't understand:
- how this root/config path interferes with TS config paths resolution
- which project should be fixed (nx/vite? vitest? code-pushup?) 

The closest issue I found in Nx reporitory is [this one].

**Comment by getlarge** [comment]
I just created [an issue in Nx](https://github.com/nrwl/nx/issues/22001) about that...

**Comment by matejchalk** [comment]
@getlarge First of all, thanks a lot for all your effort :heart: This Nx/Vitest update turned out to be a huge pain, but using all your fixes and comments saved me a lot of extra time and frustration :pray: 

I managed to eventually put together an Nx 17.3.2 update with a fully passing CI. There are a couple of failing integration tests left which in the end I decided to skip for now (ESLint plugin runner + Git helpers), so that we get something working and unblock other work.

Overall, I included most of the fixes you came up with:
- setting everything to single-thread, otherwise it would hang/timeout :+1:
  - the one downside here is it turns out the `plugin-eslint` integration tests only work with multiple threads :grimacing:  - digged into them for quite some time, but in the end decided to leave it for another day :yawning_face: 
- global setup paths are now indeed relative to project root :+1: 
- the `nxViteTsPaths()` plugin doesn't seem to do anything for Vitest anymore (also tried [vite-tsconfig-paths](https://www.npmjs.com/package/vite-tsconfig-paths) to no avail - in general I got the impression none of the `plugins` were being called) - but setting the `test.alias` worked :+1: 
  - to make it easier to maintain I created a helper function in [tools/vitest-tsconfig-path-aliases.ts](https://github.com/code-pushup/cli/blob/64e11c331ff5a9bef9d3e271581be18ec0443dac/tools/vitest-tsconfig-path-aliases.ts) so that we can just do `alias: tsconfigPathAliases()` in all vite configs - the function generates the aliases from `paths` in `tsconfig.base.json`
  - I dropped `nxViteTsPaths()` from `plugins` completely

I then ended up having to do some additional fixes, the main ones being:
- updated snapshots according to new Vitest formatting (`"`s are not escaped anymore) 
- `teardown` functions from global setup files weren't being called anymore because Nx was suddenly running `vitest` in watch mode - fixed by adding `"watch": false` to executor defaults in `nx.json`
- the coverage configuration for `examples/react-todos-app` had to be tweaked post-update, we run E2E tests against it for our `coverage-plugin`
- the `reportsDirectory` in `project.json` only was being ignored, had to copy it to vitest config for all projects
- running tests with `--coverage` would crash with `TypeError: Cannot read properties of undefined (reading 'push')` after update, seems related to https://github.com/vitest-dev/vitest/issues/4307 - fixed by using `--coverage.enabled` as suggested in issue

I have all these changes in the following PR (CI passes :relieved:):
- #532 

I think it's best to merge a working update first before proceding with `nx release` work, so that the scope doesn't get out of hand :pray: But I'm also happy to merge a PR you author instead (should also be reviewed by @Tlacenka as she's our QA), since you did the heavy lifting it would be nice to have the contributions under your name :slightly_smiling_face: Let me know your thoughts. 

And once again, huge thanks for digging into this mess :star: 

cc @BioPhoton @Tlacenka

**Comment by getlarge** [comment]
@matejchalk I am glad this investigation could make your life easier and you are done with the migration. 

> I think it's best to merge a working update first before proceding with nx release work, so that the scope doesn't get out of hand 🙏 But I'm also happy to merge a PR you author instead (should also be reviewed by @Tlacenka as she's our QA), since you did the heavy lifting it would be nice to have the contributions under your name 🙂 Let me know your thoughts.

Yes, solving this migration before going further with `nx release` is better. You can ping me once #532 is merged and i will continue #511 from the main branch.

**Comment by matejchalk** [comment]
@getlarge The update is merged and we fixed the remaining integration tests as well. So there should be no more blockers for working on `nx release` :crossed_fingers:
